### PR TITLE
Feature/allow upgrade to snapshots

### PR DESCRIPTION
**Pull request by pchila**
## What does this PR do?
This change allows to upgrade to a specific snapshot version by specifying the build id.
For example, we can now target a specific 8.8.0 snapshot build with id `49c28bdb` with 
`elastic-agent upgrade 8.8.0-SNAPSHOT+49c28bdb`

Of course the form without a build id is still valid and the agent will still query the artifact API to figure out what is the latest snapshot version available
`elastic-agent upgrade 8.8.0-SNAPSHOT`

We can have a look at the builds available for a specific snapshot by checking the artifact API at  https://artifacts-api.elastic.co/v1/versions/8.8.0-SNAPSHOT/builds/

A parsed semver version object has been introduced (and tested) in `pkg/version` (we already have at least 1 direct dependency to a semver library and an indirect one to another semver library, but since the code is small I thought it was useful to have our own version so we can implement format changes or utility methods easily)

We are using this parsed semver version in the `step_download.go` and the snapshot downloader and verifier: using this we are able to detect if we have an exact snapshot version already and we can take a shortcut through the `snapshotURI()` function of the snapshot downloader without querying the Artifact API.

For the download and the verification themselves we use the `VersionWithPrerelease()` method to calculate the correct file name that we have to download.

I left the `localremote` and the `http.Downloader` pretty much untouched so that the previous defaults are still in place and we can still upgrade using a 8.9.0-SNAPSHOT stack that specifies '8.9.0' as version (have a look at the `localremote` downloader for more info)

## Why is it important?

We want users to be able to test unreleased versions (snapshots) of elastic agent but with some control over the exact version of the software (that is we want to upgrade to a specific build)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- [x] I have added an integration test or an E2E test

## Author's Checklist

- [ ]

## Related issues

- Supersedes #2704 
- Requires #2750 
- Requires #2751 

## Questions to ask yourself

- How are we going to support this in production? 
- How are we going to measure its adoption? 
- How are we going to debug this?
- What are the metrics I should take care of?
- ...

### CODE REVIEW

File: changelog/fragments/1684731814-Allow-upgrade-to-snapshots.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+# - enhancement: extends functionality but does not break or fix existing behavior
+# - feature: new functionality
+# - known-issue: problems that we are aware of in a given version
+# - security: impacts on the security of a product or a user’s deployment.
+# - upgrade: important information for someone upgrading from a prior version
+# - other: does not fit into any of the other categories
+kind: feature
+
+# Change summary; a 80ish characters long description of the change.
+summary: Allow upgrade to snapshots
```
Comment (blakerouse) [inline changelog/fragments/1684731814-Allow-upgrade-to-snapshots.yaml:14]: This summary should really be what you have in description. I believe we always supported upgrading to a snapshot? Just not to a specific build?

File: internal/pkg/agent/application/upgrade/artifact/download/snapshot/downloader.go [hunk 2]
```diff
@@ -82,13 +88,19 @@ func snapshotConfig(config *artifact.Config, versionOverride string) (*artifact.
 	}, nil
 }
 
-func snapshotURI(versionOverride string, config *artifact.Config) (string, error) {
+func snapshotURI(versionOverride *agtversion.ParsedSemVer, config *artifact.Config) (string, error) {
+	// do we support upgrade without specifying a target version ?
```
Comment (blakerouse) [inline internal/pkg/agent/application/upgrade/artifact/download/snapshot/downloader.go:92]: What is this question here for? Does this need to be answered before this is truely ready?

File: internal/pkg/agent/application/upgrade/step_download.go [hunk 3]
```diff
@@ -47,7 +48,12 @@ func (u *Upgrader) downloadArtifact(ctx context.Context, version, sourceURI stri
...
 
-	fetcher, err := newDownloader(version, u.log, &settings)
+	parsedVersion, err := agtversion.ParseVersion(version)
+	if err != nil {
+		return "", fmt.Errorf("error parsing version %q: %w", version, err)
+	}
+
+	fetcher, err := newDownloader(parsedVersion, u.log, &settings)
 	if err != nil {
 		return "", errors.New(err, "initiating fetcher")
```
Comment (blakerouse) [inline internal/pkg/agent/application/upgrade/step_download.go:58]: While you are here, I would just go ahead and change from `errors.New(err...)` to `fmt.Errorf`. I mean that through the whole file.

File: internal/pkg/agent/application/upgrade/step_download.go [hunk 4]
```diff
@@ -65,23 +74,26 @@ func (u *Upgrader) downloadArtifact(ctx context.Context, version, sourceURI stri
...
 
-func newDownloader(version string, log *logger.Logger, settings *artifact.Config) (download.Downloader, error) {
-	if !strings.HasSuffix(version, "-SNAPSHOT") {
+func newDownloader(version *agtversion.ParsedSemVer, log *logger.Logger, settings *artifact.Config) (download.Downloader, error) {
+
+	if !version.IsSnapshot() {
 		return localremote.NewDownloader(log, settings)
 	}
 
+	// TODO since we know if it's a snapshot or not, shouldn't we add EITHER the snapshot downloader OR the release one ?
```
Comment (blakerouse) [inline internal/pkg/agent/application/upgrade/step_download.go:95]: I think so.

File: testing/integration/upgrade_test.go [hunk 5]
```diff
@@ -157,3 +163,170 @@ func getPreviousMinorVersion(v string) (string, error) {
...
+		}
+		s.T().Logf("current agent state: %+v", state)
+		return state.Info.Commit == expectedAgentHashAfterUpgrade && state.State == cproto.State_HEALTHY
+	}, 10*time.Minute, 1*time.Second, "agent never upgraded to expected version")
+
+	updateMarkerFile := filepath.Join(paths.DefaultBasePath, "Elastic", "Agent", "data", ".update-marker")
+
+	s.Require().FileExists(updateMarkerFile)
+
+	// The checks of the update marker makes the test time out since it runs for more than 10 minutes :(
```
Comment (blakerouse) [inline testing/integration/upgrade_test.go:323]: Yeah I think we need to figure out if we can shorten this somehow for out testing purposes. Waiting 10 minutes is not really something that any developer wants to wait on.

File: internal/pkg/agent/application/upgrade/step_download.go (hunk 3 above)
Comment (pchila) [reply internal/pkg/agent/application/upgrade/step_download.go:58]: Does that mean that the usage of package `github.com/elastic/elastic-agent/internal/pkg/agent/errors` is deprecated ? (I am more than happy if it is and we use the go stdlib error wrapping :tada: )

Comment (blakerouse) [reply internal/pkg/agent/application/upgrade/step_download.go:58]: Oh yes, I want it gone!

File: testing/integration/upgrade_test.go (hunk 5 above)
Comment (blakerouse) [reply testing/integration/upgrade_test.go:323]: Yeah I just wonder if we should see if we can shorten the time windows somehow just for the test.

File: internal/pkg/agent/application/upgrade/artifact/download/localremote/downloader.go [hunk 6]
```diff
@@ -21,9 +21,12 @@ func NewDownloader(log *logger.Logger, config *artifact.Config) (download.Downlo
 	downloaders := make([]download.Downloader, 0, 3)
 	downloaders = append(downloaders, fs.NewDownloader(config))
 
+	// if the current build is a snapshot we use this downloader to update
+	// to the latest snapshot of the same version useful for testing with
+	// a snapshot version of fleet for example
```
Comment (ycombinator) [inline internal/pkg/agent/application/upgrade/artifact/download/localremote/downloader.go:26]: Thanks for adding this comment.  
```suggestion
	// If the current build is a snapshot we use this downloader to update
	// to the latest snapshot of the same version. Useful for testing with
	// a snapshot version of fleet, for example.
```

File: testing/integration/upgrade_test.go (hunk 5 above)
Comment (ycombinator) [reply testing/integration/upgrade_test.go:323]: Probably not for this PR, but we could enhance the watcher to try and look for an optional `watcher.properties` or `watcher.test.properties` file.  This file could hold an override value for the `gracePeriodDuration` that's used by the watcher (as well as any other similar settings used by the error checker and crash checker).  Then, (only) tests could create this file before kicking off the upgrade.

**Review by ycombinator — APPROVED** [review]
LGTM.  

Left a couple of suggestions, one is just a minor comment cleanup the other is probably for a follow up PR.

File: testing/integration/upgrade_test.go (hunk 5 above)
Comment (pchila) [reply testing/integration/upgrade_test.go:323]: Opened https://github.com/elastic/elastic-agent/issues/2796

Comment (ycombinator) [reply testing/integration/upgrade_test.go:323]: @pchila Could you create an issue to discuss how to shorten the watcher's grace period duration and possibly other similar duration-related settings and link it from the commented out code in this PR?  That way we don't forget about this conversation here.  After that, I'm good to LGTM this PR.

Comment (ycombinator) [reply testing/integration/upgrade_test.go:323]: @pchila Looks like you created the issue just as I was typing my previous comment 🙂.  Thanks.  Just reference it from the commented out code in the test here and I'lll LGTM this PR.

**Review by ycombinator — APPROVED** [review]

**Pull request merged**